rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java

```diff
@@ -327,4 +348,36 @@ public void initializeState(FunctionInitializationContext context) throws Except
     public TypeInformation<OUT> getProducedType() {
         return schema.getProducedType();
     }
+
+    @Override
+    public void notifyCheckpointComplete(long checkpointId) throws Exception {
+        // callback when checkpoint complete
+        if (!runningChecker.isRunning()) {
+            LOG.debug("notifyCheckpointComplete() called on closed source; returning null.");
+            return;
+        }
+
+        final int posInMap = pendingOffsetsToCommit.indexOf(checkpointId);
+        if (posInMap == -1) {
+            LOG.warn("Received confirmation for unknown checkpoint id {}", checkpointId);
+            return;
+        }
+
+        Map<MessageQueue, Long> offsets = (Map<MessageQueue, Long>)pendingOffsetsToCommit.remove(posInMap);
+
+        // remove older checkpoints in map
+        for (int i = 0; i < posInMap; i++) {
+            pendingOffsetsToCommit.remove(0);
+        }
+
+        if (offsets == null || offsets.size() == 0) {
+            LOG.debug("Checkpoint state was empty.");
+            return;
+        }
+
+        for (Map.Entry<MessageQueue, Long> entry : offsets.entrySet()) {
+            consumer.updateConsumeOffset(entry.getKey(), entry.getValue());
```

Comment:
It's better to persist offsets to broker immediately in case of process crash.
>consumer.updateConsumeOffset(mq, offset);
consumer.getOffsetStore().persist(mq);

Moreover, the MQClientInstance will persist consumer offsets to broker in a fixed interval (5sec by default). See:
https://github.com/apache/rocketmq/blob/8e44ca1ff1c8a8c3513cc88087c83db04973a47a/client/src/main/java/org/apache/rocketmq/client/impl/factory/MQClientInstance.java#L301

Reply:
> It's better to persist offsets to broker immediately in case of process crash.  

If we kill the application manually. Maybe checkpoint haven't process successful. When application start again,the comsumer recover offset from broker server. In above situation,  some offset we will never re-comsume(because we persist offset immediately) . hence, we should cache the offset and  persist offsets to broker after checkpoint finish.

> It would be more reasonable to add a switch for the automatic commit and turn off the switch in the streaming environment.

In fact, variate of `enableCheckpoint` is the switch. when checkpoint is enable, rocketmq source commit offset manually; if checkpoint is disable, rocketmq source commit offset automatic(auto commit offset after process every msg) .

Reply:
@SuXingLee 
I guess you misread my comment. Please take a look again at my comment position(line number).
I mean the `consumer.updateConsumeOffset(mq, offset)` method  actually won't persist immediately.

consumer.updateConsumeOffset
->
defaultMQPullConsumerImpl.updateConsumeOffset
->
OffsetStore.updateOffset
->
update RemoteBrokerOffsetStore.offsetTable **NOT persist to broker immediately**

Reply:
@vesense you are right